### ISSUE TITLE
Analyze changes we've made to v0.45 in our fork to identify changes we need to apply to v0.46.

### ISSUE THREAD

**Issue by SpicyLemon** [provenance-io/provenance#971]
## Summary

Analyze the changes that we've made to v0.45 on our fork, with what Cosmos-SDK v0.46 contains.

## Proposal

Identify the changes that we need to re-apply to v0.46 in order to use it in Provenance.

**Comment by SpicyLemon** [comment]
Our fork has the following changes applied to it:
* root/multistore fix
* CountAuthorization
* Custom Message Fees
* Bank Send fix
* Upgrade Order fix
* Plugins
* Misc patches.

All of the Misc patches have either been applied to v0.46 or are no longer applicable.

Easy changes (probably) to re-apply:
* root/multistore fix:
   ```
   ae3b51341 fix: store/rootmulti: fix non-deterministic map iteration (#12487)
   ```
* CountAuthorization:
   ```
   a3e69b631 register CountAuthorization interface with registry (#194)
   7aa78dc13 Merge pull request #175 from provenance-io/prov/limit-authz-uses
   ```
* Upgrade Order:
   ```
   ce6dd70d5 Merge branch 'cleanup-release-pio/v0.45.x-fix-upgrade-order' into cleanup-release-pio/v0.45.x
   d3728a0b8 (origin/cleanup-release-pio/v0.45.x-fix-upgrade-order) Fix the ordering of upgrades.
   ```

More complex changes (probably) to re-apply:
* Custom Message Fees:
   ```
   0552ae84b Merge branch 'cleanup-release-pio/v0.45.x-fee-handler' into cleanup-release-pio/v0.45.x
   3dca2becb (origin/cleanup-release-pio/v0.45.x-fee-handler) Implement the Custom Message Fees.
   ```
* Bank Send:
   ```
   17f783ccd [11859]: Add a function for getting an entry so that users can differentiate between a missing entry and one that's using the default value.
   efefc29f3 Merge branch 'release-pio/v0.45.x-w-bank-send-fix' into release-pio/v0.45.x
   155e01b92 Merge branch 'release-pio/v0.45.x-w-bank-send-fix' into release-pio/v0.45.x
   f560a837d Pull in changes from dwedul/11859-send-disabled-change that move the SendEnabled info from Params to state.
   ```
* Plugins:
   ```
   5a694e62b (origin/cleanup-release-pio/v0.45.x) Merge branch 'cleanup-release-pio/v0.45.x-plugin-arch' into cleanup-release-pio/v0.45.x
   49209c59a (origin/cleanup-release-pio/v0.45.x-plugin-arch) Implement the state plugin listener stuff.
   ```

**Comment by SpicyLemon** [comment]
Here's my full investigation notes: [cosmos-our-v0.45-fork-to-v0.46_scratchpad.txt](https://github.com/provenance-io/provenance/files/9193913/cosmos-our-v0.45-fork-to-v0.46_scratchpad.txt)

**Comment by SpicyLemon** [comment]
Small update.

Turns out the upgrade order fix wasn't actually needed as the SDK had a fix for it in version 0.46.  Our fix was just to `GetLastCompletedUpgrade` but their fix was more involved. They changed how the done upgrades are stored. Previously, they were `{key}{name}` -> `{height}`. With v0.46, they are now `{key}{height}{name}` -> `1`. With that updated storage key, the reverse iterator now correctly gets the most recent upgrade first.
